docs/webactions.md

```diff
@@ -0,0 +1,141 @@
+
+1. `headers`: a JSON object where the keys are header-names and the values are string values for those headers (default is no headers).
+1. `code`: a valid HTTP status code (default is 200 OK).
+1. `body`: a string which is either plain text or a base64 encoded string (for binary data).
+
+The controller will pass along the action-specified headers, if any, to the HTTP client when terminating the request/response. Similarly the controller will respond with the given status code when present. Lastly, the body is passed along as the body of the response. Unless a `content-type header` is declared in the action result’s `headers`, the body is passed along as is if it’s a string (or results in an error otherwise). When the `content-type` is defined, the controller will determine if the response is binary data or plain text and decode the string using a base64 decoder as needed. Should the body fail to decoded correctly, an error is returned to the caller.
+
+Here is an example of a web action that performs an HTTP redirect:
+```
+function main() {
+   return { headers: { location: "http://openwhisk.org" }, 
+            code: 302 }
+}
+```  
+
+Or sets a cookie:
+```
+function main() {
+   return { headers: { "Set-Cookie": "UserID=Jane; Max-Age=3600; Version=" }, 
+            code: 200, 
+            body: "<html><body><h3>hello</h3></body></html" }
+}
+```
+
+Or returns an `image/png`:
+```
+function main() {
+    let png = <base 64 encoded string>
+    return { headers: { 'Content-Type': 'image/png' },
+             code: 200,
+             body: png };
+}
+```
+
+It is important to be aware of the [response size limit](reference.md) for actions when using `.http` response types since a response that exceeds the predefined system limits will fail. Large objects should not be sent inline through OpenWhisk, but instead deferred to an object store, for example.
+
+## HTTP Context
+
+A web action, when invoked, receives all the HTTP request information available as additional parameters to the action input argument. They are:
+
+1. `__ow_meta_verb`: the HTTP method of the request.
+2. `__ow_meta_headers`: the request headers.
+3. `__ow_meta_path`: the unmatched path of the request (matching stops after consuming the action extension).
+
+The request may not override any of the named `__ow_` parameters above; doing so will result in a failed request with status equal to 400 Bad Request.
+Action parameters may also be protected and treated as immutable. To finalize parameters, and to make an action web accessible, two [annotations](annotations.md) must be attached to the action: `final` and `web-export` either of which must be set to `true` to have affect. Revisiting the action deployment earlier, we add the annotations as follows:
+
+```
+$ wsk action create /guest/demo/hello hello.js \
+      --parameter name Jane \
+      --annotation final true \
+      --annotation web-export true
```

Comment:
Had to scroll way too far to find the example command. Maybe put it straight above below the initial example?